File: 2D Array (matrix)/54. Spiral Matrix/sprial_matrix.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans;
        ans = new ArrayList<>();

        int top = 0; //row
        int right = matrix[0].length - 1; //col
        int bottom = matrix.length - 1; //row
        int left = 0; //col

        while (top <= bottom && left <= right) {
            for (int i = top; i <= right; i++) {
                ans.add(matrix[top][i]);
            }
            top++;
            for (int i = top; i <= bottom; i++) {
                ans.add(matrix[i][right]);
            }
            right--;
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    ans.add(matrix[bottom][i]);
                }
                bottom--;
            }
            if (left <= right) {
                for (int i = bottom; i >= top; i--) {
                    ans.add(matrix[i][left]);
                }
                left++;
            }
        }
        return ans;
    }
}

public class sprial_matrix {

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[][] matrix = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };
        List<Integer> result = solution.spiralOrder(matrix);
        System.out.println(result);
    }
}

File: 2D Array (matrix)/54. Spiral Matrix/sprial_matrix.scala
object SpiralMatrix {
    def spiralOrder(matrix: Array[Array[Int]]): List[Int] = {
        if (matrix.isEmpty || matrix(0).isEmpty) return List()

        val rows = matrix.length
        val cols = matrix(0).length
        var top = 0
        var bottom = rows - 1
        var left = 0
        var right = cols - 1
        var result = List[Int]()

        while (top <= bottom && left <= right) {
            // Traverse from left to right
            for (i <- left to right) {
                result = result :+ matrix(top)(i)
            }
            top += 1

            // Traverse from top to bottom
            for (i <- top to bottom) {
                result = result :+ matrix(i)(right)
            }
            right -= 1

            // Traverse from right to left
            if (top <= bottom) {
                for (i <- right to left by -1) {
                    result = result :+ matrix(bottom)(i)
                }
                bottom -= 1
            }

            // Traverse from bottom to top
            if (left <= right) {
                for (i <- bottom to top by -1) {
                    result = result :+ matrix(i)(left)
                }
                left += 1
            }
        }

        result
    }

    def main(args: Array[String]): Unit = {
        val matrix = Array(
            Array(1, 2, 3),
            Array(4, 5, 6),
            Array(7, 8, 9)
        )

        val result = spiralOrder(matrix)
        println(result.mkString(", "))
    }
}